### PR TITLE
fix: ito swap active state; swap amount display #2225

### PR DESCRIPTION
**Pull request by lelenei**
#2225

### CODE REVIEW

**Review by septs — APPROVED** [review]

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 1]
```diff
@@ -150,6 +150,30 @@ const TokenItem = ({ price, token, exchangeToken }: TokenItemProps) => {
...
+    const _amount = formatBalance(amount, token.decimals)
+    const [number, decimal] = _amount.split('.')
+
+    if (!decimal) return number
+
+    if (number.length <= 6) {
+        if (decimal.length >= 6) {
+            return `${number}.${decimal.slice(0, 6)}`
+        }
+        return `${number}.${decimal.padEnd(6, '0')}`
```
Comment (septs) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:166]:
```ts
const balance = formatBalance(amount, token.decimals)
const sd = balance.precision(true)
return balance.toFixed(sd.length <= 12 ? clamp(sd.length, 6, 12) : 0)
```

see https://mdn.io/Math/toFixed
see https://lodash.com/docs/4.17.15#clamp
see https://mikemcl.github.io/bignumber.js/#toFix

**Review by septs — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx (hunk 1 above)
Comment (septs) [reply packages/maskbook/src/plugins/ITO/UI/ITO.tsx:166]: the sample need test and verify.

Comment (septs) [reply packages/maskbook/src/plugins/ITO/UI/ITO.tsx:166]:
```
> test = (input) => { const balance = new BigNumber(input); return (balance.precision(true) - balance.decimalPlaces()) <= 11 ? balance.toPrecision(11) : balance.toFixed(0) }
[Function: test]
> test('123456.78999')
'123456.78999'
> test('123456.789')
'123456.78900'
> test('123456.7')
'123456.70000'
> test('123456000.7')
'123456000.70'
> test('123456000000.7')
'123456000001'
```

Comment (lelenei) [reply packages/maskbook/src/plugins/ITO/UI/ITO.tsx]: ok

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 2]
```diff
@@ -151,6 +151,20 @@ const TokenItem = ({ price, token, exchangeToken }: TokenItemProps) => {
...
+    const _amount = new BigNumber(formatBalance(amount, token.decimals))
+    if (_amount.isZero()) {
+        return '0'
+    }
+    const __format = (balance: BigNumber): string => {
+        const len = balance.precision(true) - balance.decimalPlaces()
+        return len <= 6 ? balance.toPrecision(len + 6) : len >= 12 ? balance.toPrecision(len) : balance.toPrecision(12)
+    }
+
+    console.log(__format(_amount))
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:164]: Remove it.

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 3]
```diff
@@ -151,6 +151,20 @@ const TokenItem = ({ price, token, exchangeToken }: TokenItemProps) => {
     )
 }
 
+function _format(amount: BigNumber, token: EtherTokenDetailed | ERC20TokenDetailed): string {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:154]: Use a better name.

**Review by guanbinrui — DISMISSED** [review]

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 4]
```diff
@@ -291,22 +293,25 @@ export function ITO(props: ITO_Props) {
...
-                            {(listOfStatus.includes(ITO_Status.completed) && isBuyer) || total_remaining.isEqualTo(0)
-                                ? t('plugin_ito_completed')
-                                : listOfStatus.includes(ITO_Status.expired)
-                                ? t('plugin_ito_expired')
-                                : null}
-                        </Typography>
-                    ) : null}
+
+                    <Typography variant="body2" className={classes.status}>
+                        {listOfStatus.includes(ITO_Status.waited)
```
Comment (septs) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:298]: extract to function.

File: packages/maskbook/src/plugins/Wallet/formatter.ts [hunk 5]
```diff
@@ -106,3 +106,23 @@ export function formatElapsed(from: number) {
...
+    const _amount = new BigNumber(amount)
+    const _decimalPlaces = decimalPlaces < 0 ? 6 : decimalPlaces
+    const _precision = precision < 0 ? 12 : precision
+    if (_amount.isZero()) {
+        return '0'
+    }
+
+    const __format = (balance: BigNumber): string => {
+        const len = balance.precision(true) - balance.decimalPlaces()
+        return len <= _decimalPlaces
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/formatter.ts:120]: plz do not use ternary expr.

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx (hunk 4 above)
Comment (septs) [reply packages/maskbook/src/plugins/ITO/UI/ITO.tsx:298]: plz do not use ternary expr.

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 6]
```diff
@@ -291,22 +293,25 @@ export function ITO(props: ITO_Props) {
...
+                            ? t('plugin_ito_status_no_start')
+                            : listOfStatus.includes(ITO_Status.expired)
+                            ? t('plugin_ito_expired')
+                            : listOfStatus.includes(ITO_Status.started)
+                            ? t('plugin_ito_status_ongoing')
+                            : listOfStatus.includes(ITO_Status.started) && total_remaining.isEqualTo(0)
+                            ? t('plugin_ito_status_out_of_stock')
+                            : (listOfStatus.includes(ITO_Status.completed) && isBuyer) || total_remaining.isEqualTo(0)
```
Comment (jk234ert) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:306]: Too complex in one expr

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx (hunk 1 above)
Comment (lelenei) [reply packages/maskbook/src/plugins/ITO/UI/ITO.tsx]: balance.precision() ： Number of significant digits

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 7]
```diff
@@ -291,22 +293,25 @@ export function ITO(props: ITO_Props) {
...
-                    ) : null}
+
+                    <Typography variant="body2" className={classes.status}>
+                        {listOfStatus.includes(ITO_Status.waited)
+                            ? t('plugin_ito_status_no_start')
+                            : listOfStatus.includes(ITO_Status.expired)
+                            ? t('plugin_ito_expired')
+                            : listOfStatus.includes(ITO_Status.started)
+                            ? t('plugin_ito_status_ongoing')
+                            : listOfStatus.includes(ITO_Status.started) && total_remaining.isEqualTo(0)
```
Comment (septs) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:304]:
```suggestion
                            : listOfStatus.includes(ITO_Status.started) && total_remaining.isZero()
```

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 8]
```diff
@@ -374,7 +401,9 @@ export function ITO(props: ITO_Props) {
                                     {listOfStatus.includes(ITO_Status.waited)
                                         ? `Start date: ${formatDateTime(new Date(start_time * 1000), true)}`
```
Comment (septs) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:402]: why not i18n

**Review by Jack-Works — APPROVED** [review]

File: packages/maskbook/src/plugins/Wallet/formatter.ts [hunk 9]
```diff
@@ -106,3 +106,24 @@ export function formatElapsed(from: number) {
...
+    if (_amount.isZero()) {
+        return '0'
+    }
+
+    if (_amount.isLessThan(1)) {
+        return _amount.toFixed(_precision)
+    }
+
+    const len = _amount.precision() - _amount.decimalPlaces()
+    return len <= _decimalPlaces
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/formatter.ts:124]: plz expand the expr

File: packages/maskbook/src/plugins/ITO/UI/ConfirmDialog.tsx [hunk 10]
```diff
@@ -223,9 +223,11 @@ export function ConfirmDialog(props: ConfirmDialogProps) {
                 <Grid item xs={6} className={classes.button}>
                     <ActionButton fullWidth variant="contained" onClick={onDone}>
                         {t('plugin_ito_send_text', {
-                            total: formatBalance(
-                                new BigNumber(poolSettings?.total ?? '0'),
-                                poolSettings?.token?.decimals ?? 0,
+                            total: formatAmountPrecision(
+                                formatBalance(
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/ITO/UI/ConfirmDialog.tsx:227]: Invoke `formatBalance` in the `formatAmountPrecision`?

File: packages/maskbook/src/plugins/ITO/UI/ITO.tsx [hunk 11]
```diff
@@ -284,29 +286,116 @@ export function ITO(props: ITO_Props) {
...
+        () => (
+            <>
+                <Typography variant="body1">{swapResultText}</Typography>
+                {footerEndTime}
+            </>
+        ),
+        [footerEndTime, swapResultText],
+    )
+
+    const footerNormal = useMemo(
```
Comment (septs) [inline packages/maskbook/src/plugins/ITO/UI/ITO.tsx:363]: why not design as component?

Comment (lelenei) [reply packages/maskbook/src/plugins/ITO/UI/ITO.tsx:363]: I think it's the same

Comment (lelenei) [reply packages/maskbook/src/plugins/ITO/UI/ITO.tsx:363]: This code is simple

**Review by Jack-Works — APPROVED** [review]

**Review by jk234ert — APPROVED** [review]

**Pull request merged**